File: src/main/java/io/github/bot/terminal/application/domains/db_connection/entity/DbConnection.java
package io.github.bot.terminal.application.domains.db_connection.entity;

import io.github.bot.terminal.application.domains.common.Persistable;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.datasource.DriverManagerDataSource;

import javax.sql.DataSource;
import java.util.Optional;

public abstract class DbConnection<D extends DbConnectionDetails> implements Persistable {

    public abstract D getDetails();

    public abstract String getUrl();

    public abstract String getClassName();

    public abstract String getCheckSelect();

    public void check() {
        DriverManagerDataSource dataSource = new DriverManagerDataSource();
        dataSource.setDriverClassName(getClassName());
        dataSource.setUrl(getUrl());
        dataSource.setUsername(getDetails().getUser());
        dataSource.setPassword(getDetails().getPassword());
        JdbcTemplate template = new JdbcTemplate(dataSource);
        template.execute(getCheckSelect());
    }

    public Optional<DataSource> getDataSource(String dbConnectionId) {
        try {
            DriverManagerDataSource dataSource = new DriverManagerDataSource();
            dataSource.setDriverClassName(getClassName());
            dataSource.setUrl(getUrl());
            dataSource.setUsername(getDetails().getUser());
            dataSource.setPassword(getDetails().getPassword());
            return Optional.of(dataSource);
        } catch (Exception ex) {
            return Optional.empty();
        }
    }
}

File: src/main/kotlin/io/github/support/terminal/application/domains/db_connection/entity/DbConnection.kt
package io.github.support.terminal.application.domains.db_connection.entity

import io.github.support.terminal.application.domains.db_connection.repository.DbConnectionDetails
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.datasource.DriverManagerDataSource
import java.util.*
import javax.sql.DataSource

abstract class DbConnection<D : DbConnectionDetails> {

    abstract val details: D
    abstract val id: String
    abstract val url: String
    abstract val className: String
    abstract val checkSelect: String

    fun check() {
        val template = createJdbcTemplate()
        template.execute(checkSelect)
    }

    fun createJdbcTemplate(): JdbcTemplate {
        return JdbcTemplate(dataSource.get())
    }

    //TODO log out -- !!!When class name not found - it means that we forget jar driver or we dont see it
    val dataSource: Optional<DataSource>
        get() = try {
            val dataSource = DriverManagerDataSource()
            dataSource.url = url
            dataSource.username = details.user
            dataSource.password = details.password
            dataSource.setDriverClassName(className)
            Optional.of(dataSource)
        } catch (ex: Exception) {
            //TODO log out -- !!!When class name not found - it means that we forget jar driver or we dont see it
            Optional.empty()
        }
}